src/librustc_typeck/check/demand.rs

```diff
@@ -88,10 +94,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     pub fn demand_coerce_diag(&self,
                               expr: &hir::Expr,
                               checked_ty: Ty<'tcx>,
-                              expected: Ty<'tcx>) -> Option<DiagnosticBuilder<'tcx>> {
+                              expected: Ty<'tcx>)
+                              -> (Ty<'tcx>, Option<DiagnosticBuilder<'tcx>>) {
         let expected = self.resolve_type_vars_with_obligations(expected);
 
-        if let Err(e) = self.try_coerce(expr, checked_ty, self.diverges.get(), expected) {
+        self.try_coerce(expr, checked_ty, self.diverges.get(), expected).map_err(|e| {
```

Comment:
Nit: this whole `foo.map_err().map().unwrap_or` is getting a bit over the top and kind of hard to follow. Can we just rewrite this with a `match`?

Something like:

```rust
let e = match self.try_coerce(...) {
    Ok(ty) => return (ty, None),
    Err(e) => e,
};

...

(expected, some(err))
```

Reply:
I wanted to avoid intending the whole function right, but I'm okay with a match either.